### PR TITLE
core utils implementation

### PR DESCRIPTION
**Pull request by Nabil-Salah**
The following coreutils are implemented:

head
tail
wc (word count)
cat
echo
tree
env
true
false

with the Acceptance Criteria of:

Each utility is implemented as a separate Go program.
Includes basic error handling.
Includes a README with build and usage instructions for all utilities.

### CODE REVIEW

File: cmd/cat/main.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	if e != nil {
+		log.Fatal("Incorrect file path or file does not exist ", e)
+	}
+}
+
+func main() {
+	var n bool
+	flag.BoolVar(&n, "n", false, "number of lines")
+	flag.Parse()
+	if len(os.Args) == 0 {
```
Comment (Eslam-Nawara) [inline cmd/cat/main.go]: os.args[0] is always the executable you are running, If you need to make sure the user passed an argument you need to check if the length is < 2

File: cmd/cat/main.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+)
+
+func checkFile(e error) {
```
Comment (Eslam-Nawara) [inline cmd/cat/main.go]: should this function be named checkFile? it has nothing to do with files

File: cmd/cat/main.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		log.Fatal("No file passed")
+	}
+	var filepath string = os.Args[len(os.Args)-1]
+	file, err := os.Open(filepath)
+	checkFile(err)
+	defer file.Close()
+	in := bufio.NewScanner(file)
+	var cnt int = 0
+	for in.Scan() {
+		if n {
```
Comment (Eslam-Nawara) [inline cmd/cat/main.go]: you should use more descriptive names, I shouldn't need to go back and check the code to know what `n` is

File: cmd/env/main.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+func main() {
+	for _, e := range os.Environ() {
+		idx := strings.Index(e, "=")
```
Comment (Eslam-Nawara) [inline cmd/env/main.go]: why do you split the value then concatenate to print them again

File: cmd/head/main.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	in := bufio.NewScanner(file)
+	var cnt int = 0
+	for in.Scan() {
+		fmt.Println(in.Text())
+		cnt++
+		if cnt == (n) {
+			break
+		}
+	}
+	if cnt != n {
```
Comment (Eslam-Nawara) [inline cmd/head/main.go]: this shouldn't be fatal, running head with a file has less number then requested shouldn't fail, just print the lines and return

File: cmd/head/main.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	in := bufio.NewScanner(file)
+	var cnt int = 0
+	for in.Scan() {
+		fmt.Println(in.Text())
+		cnt++
+		if cnt == (numberOfLines) {
+			break
+		}
+	}
+	if cnt != numberOfLines {
```
Comment (Eslam-Nawara) [inline cmd/head/main.go]: this shouldn't be fatal, running head with a file has less number then requested shouldn't fail, just print the lines and return

File: cmd/tail/main.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	}
+}
+
+func main() {
+	var numberOfLines int
+	flag.IntVar(&numberOfLines, "n", 10, "number of lines")
+	flag.Parse()
+	if len(os.Args) <= 1 {
+		log.Fatal("No file passed")
+		os.Exit(1)
```
Comment (Eslam-Nawara) [inline cmd/tail/main.go]: no need for that

File: cmd/tail/main.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	checkFileError(err)
+	defer file.Close()
+	in := bufio.NewScanner(file)
+	var lines []string
+	for in.Scan() {
+		lines = append(lines, in.Text())
+	}
+	if (numberOfLines) > len(lines) {
+		log.Fatal("Lines Aren't enough")
+	}
```
Comment (Eslam-Nawara) [inline cmd/tail/main.go]: this shouldn't be fatal

File: cmd/wc/main.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	if printLineCount {
+		fmt.Println(lineCount)
+	}
+	if printWordCount {
+		fmt.Println(wordCount)
+	}
+	if printCharactersCount {
+		fmt.Println(charactersCount)
+	}
```
Comment (Eslam-Nawara) [inline cmd/wc/main.go]: if all are false you should print them all

**Review by Eslam-Nawara — CHANGES_REQUESTED** [review]

File: cmd/head/main.go (hunk 6 above)
Comment (Nabil-Salah) [reply cmd/head/main.go]: ok done

File: cmd/tail/main.go (hunk 7 above)
Comment (Nabil-Salah) [reply cmd/tail/main.go]: ok removed

File: cmd/tail/main.go (hunk 8 above)
Comment (Nabil-Salah) [reply cmd/tail/main.go]: done

File: cmd/wc/main.go (hunk 9 above)
Comment (Nabil-Salah) [reply cmd/wc/main.go]: done

**Review by Eslam-Nawara — APPROVED** [review]

**Pull request merged**